features/26-mfa/api.md

```diff
@@ -0,0 +1,1030 @@
+  channel: "email";
+  email: string;
+  authenticationSessionToken?: string;
+}
+
+interface TriggerOOBOptions {
+  authenticatorID?: string;
+  authenticationSessionToken?: string;
+}
+
+interface ActivateOOBOptions {
+  authenticatorID: string;
+  code: string;
+  authenticationSessionToken?: string;
+}
+
+interface CreateNewOOBResult {
+  authenticatorID: string;
+  authenticatorType: "oob";
+  channel: "sms" | "email";
+}
+
+interface ActivateOOBResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewOOB(options: CreateNewOOBOptions): Promise<CreateNewOOBResult>;
+async function triggerOOB(options?: TriggerOOBOptions): Promise<void>;
+async function activateOOB(opitons: ActivateOOBOptions): Promise<ActivateOOBResult>;
+
+// Error inspection
+
+interface AuthenticationSession {
+  token: string;
+  step: "identity" | "mfa";
+}
+
+function decodeAuthenticationSession(value: unknown): AuthenticationSession | undefined;
+
+// Authenticate
+
+interface AuthenticateWithTOTPOptions {
+  otp: string;
+  authenticationSessionToken: string;
+  skipMFAForCurrentDevice?: boolean;
+}
+
+interface AuthenticateWithOOBOptions {
+  code: string;
+  authenticationSessionToken: string;
+  skipMFAForCurrentDevice?: boolean;
+}
+
+interface AuthenticateWithRecoveryCodeOptions {
+  code: string;
+  authenticationSessionToken: string;
+};
+
+async function authenticateWithTOTP(options: AuthenticateWithTOTPOptions): Promise<User>;
+
+async function authenticateWithOOB(options: AuthenticateWithOOBOptions): Promise<User>;
+
+async function authenticateWithRecoveryCode(options: AuthenticateWithRecoveryCodeOptions): Promise<User>;
+
+// Authenticator management
+
+type Authenticator = TOTPAuthenticator | OOBSMSAuthenticator | OOBEmailAuthenticator;
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBSMSAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  maskedPhone: string;
+}
+
+interface OOBEmailAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "email";
+  maskedEmail: string;
+}
+
+interface RegenerateRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+interface ListRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+interface GetAuthenticatorsOptions {
+  authenticationSessionToken?: string;
+}
+
+async function getAuthenticators(options?: GetAuthenticatorsOptions): Promise<Authenticator[]>;
+async function deleteAuthenticator(authenticatorID: string): Promise<void>;
+async function regenerateRecoveryCode(): Promise<RegenerateRecoveryCodeResult>;
+async function listRecoveryCode(): Promise<ListRecoveryCodeResult>;
+async function revokeAllTrustedDevices(): Promise<void>;
+```
+
+# Use Cases
+
+## Registration
+
+### SDK
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  const authnSession = skygear.decodeAuthenticationSession(e);
```

Comment:
I have removed `decodeAuthenticationSession` and instead store the authentication session in commit https://github.com/SkygearIO/features/pull/343/commits/0f556dbd6f65dedf6f42400e4c23b07573aa5e8e

The API is different from your suggestion because we do not have the convention of having a dedicated error class for each kind of error.

Reply:
Sorry, I didn't make it clear... my comment was more about, can we make `authenticationSession` transparent to developer users?

Right now they still need to understand the concept and use it from `skygear.auth.authenticationSession`?

The new code sample segment below make me feel really strange:

```
try {
  await skygear.auth.login("user@example.com", "password");
} catch (e) {
  const authnSession = skygear.auth.authenticationSession;
  if (authnSession != null && authnSession.step === "mfa") {
    navigateToRegisterMFAScreen(authnSession);
    return;
  }
  // Handle any other error.
}
```

The MFA login is triggered by `try catch`, but the hints of it is MFA related is totally unrelated with the exception?

So what is the convention of how we return each kind of error? Can we just follow that convention?

Apparently, we really have no reason why we need to make the user aware of the `authenticationSession`...

Reply:
Please see the comment https://github.com/SkygearIO/features/pull/343#discussion_r324970903